### PR TITLE
Fix 64-bit int types in `type_dispatcher.cuh`

### LINKED ISSUE #4121: Incorrect c type used for 64 bit integer, Windows (effects sort, unique and presumably various other functions)
* Conditions
Windows 10

CuPy Version          : 8.0.0
CUDA Root             : C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0
CUDA Build Version    : 11000
CUDA Driver Version   : 11010
CUDA Runtime Version  : 11000
cuBLAS Version        : 11200
cuFFT Version         : 10201
cuRAND Version        : 10201
cuSOLVER Version      : (10, 6, 0)
cuSPARSE Version      : 11101
NVRTC Version         : (11, 0)
Thrust Version        : 100909
CUB Build Version     : 100909
cuDNN Build Version   : None
cuDNN Version         : None
NCCL Build Version    : None
NCCL Runtime Version  : None
cuTENSOR Version      : None

* Code to reproduce

`
A = cp.arange(2,0,-1, dtype=np.uint64)

print(A.view(np.uint8).reshape(-1,8))

cp.cuda.thrust.sort(A.dtype, A.data.ptr, 0, A.shape)

print(A.view(np.uint8).reshape(-1,8))`

expected output:
[[2 0 0 0 0 0 0 0]
 [1 0 0 0 0 0 0 0]]
[[1 0 0 0 0 0 0 0]
 [2 0 0 0 0 0 0 0]]

actual output:
[[2 0 0 0 0 0 0 0]
 [1 0 0 0 0 0 0 0]]
[[0 0 0 0 2 0 0 0]
 [1 0 0 0 0 0 0 0]]

dtype_dispatcher uses long for np.int64 and unsigned long for np.uint64. These datatypes are only guaranteed to be of 32 bit width. In unix systems these are 64 bit in length but in windows, only 32 bit in length which ends up giving the unexpected output above.

I get that Windows compatibility is not a priority but as far as I can tell, nothing would be lost in using long long/unsigned long long for windows compatibility.

This is the first issue I've raised so apologies for any oversight.

### PR DESCRIPTION
**Pull request by leofang**
Close #4121. Thanks @JamesWhalley for reporting the failure on Windows!

### CODE REVIEW

**Comment by leofang** [comment]
Jenkins, test this please

File: cupy/core/include/cupy/type_dispatcher.cuh [hunk 1]
```diff
@@ -40,11 +40,11 @@ void dtype_dispatcher(int dtype_id, functor_t f, Ts&&... args)
     case CUPY_TYPE_INT8:       return f.template operator()<char>(std::forward<Ts>(args)...);
     case CUPY_TYPE_INT16:      return f.template operator()<short>(std::forward<Ts>(args)...);
     case CUPY_TYPE_INT32:      return f.template operator()<int>(std::forward<Ts>(args)...);
-    case CUPY_TYPE_INT64:      return f.template operator()<long>(std::forward<Ts>(args)...);
+    case CUPY_TYPE_INT64:      return f.template operator()<int64_t>(std::forward<Ts>(args)...);
```
Comment (asi1024) [inline cupy/core/include/cupy/type_dispatcher.cuh:43]: The header <cstdint> will be required.

Comment (leofang) [reply cupy/core/include/cupy/type_dispatcher.cuh:43]: Yeah I was puzzled about this...All checks have passed, do you know why?

Comment (leofang) [reply cupy/core/include/cupy/type_dispatcher.cuh:43]: I noticed both `cupy/cuda/thrust.cpp` and `cupy/cuda/cub.cpp` include `stdint.h` for some reason, perhaps exposed through other `cimport`?

**Comment by leofang** [comment]
Jenkins, test this please

File: cupy/core/include/cupy/type_dispatcher.cuh (hunk 1 above)
Comment (leofang) [reply cupy/core/include/cupy/type_dispatcher.cuh:44]: Done in d760b14.

**Comment by asi1024** [comment]
LGTM!

**Pull request merged**